doc/design/request_response_communication.md

```diff
@@ -0,0 +1,118 @@
+# Request/Response Communication
+
+## Summary
+
+Beside the publish-subscribe messaging pattern initially supported by iceoryx,
+the request-response pattern is also widely used in the field where iceoryx is headed.
+Since we aim for a good integration with ROS 2, this is also necessary to support its
+`Service` communication method which is often used to control robots.
+
+While publish-subscribe is a data distribution pattern which continuously delivers
+updates on its data, request-response is a remote procedure call and task distribution pattern
+which delivers data only on demand. This means there is one or several clients
+requesting a server to perform a task.
+
+In iceoryx, the client and server will be built on top of the same building blocks
+we already use for publish-subscribe, like the `ChunkSender` and `ChunkReceiver`.
+
+## Terminology
+
+| Name              | Description                                              |
+| :---------------- | :------------------------------------------------------- |
+| Request-Response  | communication pattern which delivers data on demand      |
+| Publish-Subscribe | communication pattern which continuously delivers updates on its data |
+| RPC               | a remote procedure call to offload task to a server      |
+| Client            | an entity invoking RPCs                                  |
+| Server            | an entity processing RPCs                                |
+
+## Design
+
+### Considerations
+
+Client and server shall be able to be used in combination with the `WaitSet` and `Listener`
+to provide a means of event based communication instead of polling.
+
+Like with publish-subscribe, the `ServiceDescription` will be used to connect clients with a corresponding server.
+
+In order to support asynchronous requests, a sequence ID should be part of each request which will be copied to the corresponding response.
+
+### Solution
+
+This is an overview of the untyped `Client` and `Server` classes.
+
+![simple class diagram](diagrams/request_response/overview_class.svg)
+
+The `Client` and `Server` are reusing the `ChunkSender` and `ChunkReceiver` building blocks. The `Client` uses a `ChunkSender` to send requests and a `ChunkReceiver` to get the responses while the `Server` uses a `ChunkReceiver` to get the requests and a `ChunkSender` to send the responses.
+
+#### Typed API
+
+![typed API](diagrams/request_response/typed_api.svg)
+
+Since the `Response` is tied to a specific `Request`, the `loan` method takes a `Request` to populate the `Response` with the correct settings.
```

Comment:
I could think about some slightly different alternative mechanism as follows if the response type is known to the client, this would work more like a future (and since we are using asynchronous RPC this may be useful). Both methods do not really exclude each other.
1) Client has created request data D (arguments for the RPC call basically).
2) Client wraps this in some kind of Request-Response structure RR, which already contains room for the response and enough information for the server about what to do with the arguments (i.e. the call information)
3) Client sends RR to Server, but holds on to a reference (chunk) of it
4) Server extracts the arguments from RR and computes the response (or rejects the request, which is also some kind of response), populates the response area of RR and marks the response as ready
4) Client waits until the response is ready, by observing RR (actively or passively) - non-blocking, it can do other things in the meantime
5) Client extracts actual response information once the response is ready and releases RR

Note that this would not directly work outside of shared memory communication, where the response as to be actually send somewhere. API-wise it would still be possible though, just with sending the response under behind the scenes.